### ISSUE TITLE
DoubleNegation w/allowed_in_returns incorrectly reports offenses inside returned conditional clauses

### ISSUE THREAD

**Issue by nevans** [rubocop/rubocop#10474]
Assuming the default configuration for `Style/DoubleNegation`, where `EnforcedStyle: allowed_in_returns`...

## Expected behavior

Double negation should be valid inside hash literals or array literals which are used as return values.

_(ideally, I'd also prefer a more lax `EnforcedStyle: disallow_in_conditionals`, but that needs a feature ticket...)_

## Actual behavior and steps to reproduce the problem

with example.rb:
```ruby
def bangbang_inside_a_hash
  { foo: etc, bar: !!baz } # => okay
end

def bangbang_inside_an_array
  [!!foo0, bar0, baz0]  # => okay
end

def bangbang_inside_an_array_inside_an_if_elsif_else
  if foo?
    [!!foo0, bar0, baz0]  # => incorrectly triggers Style/DoubleNegation
  elsif bar?
    [foo1, !!bar1, baz1]  # => incorrectly triggers Style/DoubleNegation
  elsif baz?
    [foo2, bar2, !!baz2]  # => incorrectly triggers Style/DoubleNegation
  else
    [foo3, !!bar3, baz3]  # => but this is okay
  end
end

def bangbang_inside_an_if_elsif_else_inside_a_hash
  if foo?
    { foo: !!foo0, bar: bar0, baz: baz0 }  # => incorrectly triggers Style/DoubleNegation
  elsif bar?
    { foo: !!foo1, bar: bar1, baz: baz1 }  # => incorrectly triggers Style/DoubleNegation
  elsif baz?
    { foo: !!foo2, bar: bar2, baz: baz2 }  # => incorrectly triggers Style/DoubleNegation
  else
    { foo: !!foo3, bar: bar3, baz: baz3 }  # => but this is okay
  end
end

def bangbang_inside_an_case_when_inside_a_hash
  case foobar
  when Foo
    { foo: !!foo0, bar: bar0, baz: baz0 }  # => incorrectly triggers Style/DoubleNegation
  when Boo
    { foo: !!foo1, bar: bar1, baz: baz1 }  # => incorrectly triggers Style/DoubleNegation
  when FooBar
    { foo: !!foo2, bar: bar2, baz: baz2 }  # => incorrectly triggers Style/DoubleNegation
  else
    { foo: !!foo3, bar: bar3, baz: baz3 }  # => but this is okay
  end
end

def bangbang_inside_an_case_when_inside_a_hash
  case foobar
  in Foo
    { foo: !!foo0, bar: bar0, baz: baz0 }  # => incorrectly triggers Style/DoubleNegation
  in Boo
    { foo: !!foo1, bar: bar1, baz: baz1 }  # => incorrectly triggers Style/DoubleNegation
  in FooBar
    { foo: !!foo2, bar: bar2, baz: baz2 }  # => incorrectly triggers Style/DoubleNegation
  else
    { foo: !!foo3, bar: bar3, baz: baz3 }  # => but this is okay
  end
end

```

The following command:
```sh
rubocop --only Style/DoubleNegation --force-default-config --cache false --debug example.rb
```

Returns the following offenses:

```
Default configuration from /home/nick/.rbenv/versions/2.7.4/lib/ruby/gems/2.7.0/gems/rubocop-1.26.1/config/default.yml
Inspecting 1 file
Scanning /home/nick/src/mailstrom/tmp/example.rb
C

Offenses:

example.rb:13:6: C: [Correctable] Style/DoubleNegation: Avoid the use of double negation (!!).
    [!!foo0, bar0, baz0]  # => incorrectly triggers Style/DoubleNegation
     ^
example.rb:15:12: C: [Correctable] Style/DoubleNegation: Avoid the use of double negation (!!).
    [foo1, !!bar1, baz1]  # => incorrectly triggers Style/DoubleNegation
           ^
example.rb:17:18: C: [Correctable] Style/DoubleNegation: Avoid the use of double negation (!!).
    [foo2, bar2, !!baz2]  # => incorrectly triggers Style/DoubleNegation
                 ^
example.rb:25:12: C: [Correctable] Style/DoubleNegation: Avoid the use of double negation (!!).
    { foo: !!foo0, bar: bar0, baz: baz0 }  # => incorrectly triggers Style/DoubleNegation
           ^
example.rb:27:12: C: [Correctable] Style/DoubleNegation: Avoid the use of double negation (!!).
    { foo: !!foo1, bar: bar1, baz: baz1 }  # => incorrectly triggers Style/DoubleNegation
           ^
example.rb:29:12: C: [Correctable] Style/DoubleNegation: Avoid the use of double negation (!!).
    { foo: !!foo2, bar: bar2, baz: baz2 }  # => incorrectly triggers Style/DoubleNegation
           ^
example.rb:38:12: C: [Correctable] Style/DoubleNegation: Avoid the use of double negation (!!).
    { foo: !!foo0, bar: bar0, baz: baz0 }  # => incorrectly triggers Style/DoubleNegation
           ^
example.rb:40:12: C: [Correctable] Style/DoubleNegation: Avoid the use of double negation (!!).
    { foo: !!foo1, bar: bar1, baz: baz1 }  # => incorrectly triggers Style/DoubleNegation
           ^
example.rb:42:12: C: [Correctable] Style/DoubleNegation: Avoid the use of double negation (!!).
    { foo: !!foo2, bar: bar2, baz: baz2 }  # => incorrectly triggers Style/DoubleNegation
           ^
example.rb:51:12: C: [Correctable] Style/DoubleNegation: Avoid the use of double negation (!!).
    { foo: !!foo0, bar: bar0, baz: baz0 }  # => incorrectly triggers Style/DoubleNegation
           ^
example.rb:53:12: C: [Correctable] Style/DoubleNegation: Avoid the use of double negation (!!).
    { foo: !!foo1, bar: bar1, baz: baz1 }  # => incorrectly triggers Style/DoubleNegation
           ^
example.rb:55:12: C: [Correctable] Style/DoubleNegation: Avoid the use of double negation (!!).
    { foo: !!foo2, bar: bar2, baz: baz2 }  # => incorrectly triggers Style/DoubleNegation
           ^

1 file inspected, 12 offenses detected, 12 offenses auto-correctable
```

## RuboCop version
```
$ rubocop --force-default-config -V
1.26.1 (using Parser 3.1.1.0, rubocop-ast 1.16.0, running on ruby 2.7.4 x86_64-linux)
```
